Change for the condition when refund button is available

kidinov opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #6641
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
The PR adds a condition that the refund button is visible only when an order is partially or fully paid

@anitaa1990 @hichamboushaba asking you to review this too, because this changes the logic which you were working with, so we might be missing something here

### Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->
Try the following:
* Order is not paid -> NOT visible
* Order is fully paid -> Visible 
* Order is partially refunded -> Visible
* Order is fully refunded -> NOT Visible

~~Still in consultations with @joshheald about the state when the order is partially paid. I don't know how to end up in this state, therefore do not merge label~~

Apparently we don't support this at the moment in the app anyway, so we ignore this case for now


### Images/gif
<!-- Include before and after images or gifs when appropriate. -->
#### Partial Refund
<img src="https://user-images.githubusercontent.com/4923871/172364021-3a8fb06f-3ccc-4b86-a0e2-2a2fcd64a326.jpg" width="200px" />

#### Fully Paid
<img src="https://user-images.githubusercontent.com/4923871/172364028-aa26f228-5218-4f2b-ab2d-29656fccec49.jpg" width="200px" />


#### Not paid
<img src="https://user-images.githubusercontent.com/4923871/172364033-4616502c-d8c7-4ea8-84a8-6805a8b7ef6a.jpg" width="200px" />

#### Fully Refund
<img src="https://user-images.githubusercontent.com/4923871/172364039-8045fb94-b3b5-43f2-a5b6-22bf257ebd7e.jpg" width="200px" />
### Ensure that the Issue Refund button shows for orders with only Shipping lines (left) to refund
<img width="200px" alt="image" src="https://user-images.githubusercontent.com/4923871/172366096-ea41e107-6b93-4f3b-a58c-e56835728623.png">

kidinov committed:
Change for the condition when refund is available. Now the order has to be at least partially paid in order to be refundable

malinajirka reviewed (commented):
Thanks for implementing this logic @kidinov. This new behavior finally makes sense to me :).

> > Still in consultations with @joshheald about the state when the order is partially paid. I don't know how to end up in this state, therefore do not merge label

> Apparently we don't support this at the moment in the app anyway, so we ignore this case for now

👍 If I recall correctly from my support rotation, partially payments are supported only with a separate extension.

-----------

One possible issue:
1. Open an order with "Pending payment" status (unpaid).
2. Change it's status to "Completed".
3. Notice the "Collect Payment" button disappears but the "Issue refund" button remains hidden.

------------

P.S. We might want to consider creating a test suite in a follow up PR? The logic is getting complicated over time and the relations between the fields could bite us if we don't test it properly. Wdyt?

kidinov committed:
Reload order details after successful status update

kidinov committed:
Some tests on order model

kidinov commented:
@malinajirka thank you for the review

> One possible issue:

Fixed it here, not sure if it's a good fix. Apparently optimistic update based on the assumption that status change does not have any side effects on the order model, but this is not true, e.g. `complete` makes order `paid` on the backend. @wzieba could you please verify if this is ok?
1a01ca8637c3bc21a82ad56f406edce86e4cea4b

> P.S. We might want to consider creating a test suite in a follow up PR? The logic is getting complicated over time and the relations between the fields could bite us if we don't test it properly. Wdyt?

Added some tests here 4fe577622d0b6d111c7c2e8d16a4b6913d93d78b

kidinov committed:
Fixed checkstyle

malinajirka reviewed (approved):
Thanks so much for updating the code and adding the tests. I'm really sorry the review took so long. Looks good to me.


I'll wait for @wzieba's double check just to make sure we are not missing any context.

wzieba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/details/OrderDetailViewModel.kt:
```diff
@@ -468,6 +468,7 @@ final class OrderDetailViewModel @Inject constructor(
                                         prepareTracksEventsDetails(result.event)
                                     )
                                 } else {
+                                    reloadOrderDetails()
```
As `reloadOrderDetails()` happens regardless of any conditions, I suggest moving this outside `when` to increase readability.

kidinov replied:
Good point! Done here bb260f516848708711309869f9ff337d35059939

wzieba reviewed (approved):
> Apparently optimistic update based on the assumption that status change does not have any side effects on the order model, but this is not true, e.g. complete makes order paid on the backend.

Good point @kidinov - this scenario is an argument against optimistic updates in our app. As we can't determine the correct state from only local changes, we're prone to showing an invalid state :(.

Codewise it looks good - I have one comment.

kidinov committed:
reloadOrderDetails with any result
